rules/aip0158/response_plural_first_field.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0158
+
+import (
+	"fmt"
+
+	"github.com/gertd/go-pluralize"
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/jhump/protoreflect/desc"
+)
+
+var responsePluralFirstField = &lint.MessageRule{
+	Name:   lint.NewRuleName(158, "response-plural-first-field"),
+	OnlyIf: isPaginatedResponseMessage,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		// Rule check: for all messages that end in Response and contain a next_page_token field.
+		// Throw a linter warning if, the first field in the message is not named according to plural(message_name.to_snake().split('_')[1:-1]).
+		nextPageToken := m.FindFieldByName("next_page_token")
+		if nextPageToken == nil {
```

Comment:
Does line 28 not cover this?

Reply:
No, it doesn't cover. line 28 just check that pagination response matches with this > regexp.MustCompile("^(List|Search)[A-Za-z0-9]*Response$").

Reply:
Does it matter whether we include pagination or not? Just because we don't have a paginated response doesn't mean we shouldn't check whether the field here is a repeated list of resources with a pluralized name, right?

Reply:
Well, my intent was to attach this rule to the _pagination_ AIP.
I think that is still what I want to do -- it is pagination that triggers the need for this check, and if pagination is implemented, it does not matter whether the request is a list, a search, or custom.

Reply:
Got it. So, looks like now we have 2 ways to defined pagination message.
1. Message that match with ("^(List|Search)[A-Za-z0-9]*Response$").
2. Message that match with  ("[A-Za-z0-9]*Response$") and has next_page_token field.

Reply:
@lukesneeringer Why wouldn't we attach it to the List method AIP ? (https://aip.dev/132#response-message) The List AIP includes a pointer to requirements stemming from the Pagination AIP, and this particular field that we care about here is relevant even in cases where an API decides not to implement pagination (for whatever reason).

Reply:
We don't have a rule for the Search method. Should we have a new rule for this? or should I update these functions?
1. [isPaginatedResponseMessage](https://github.com/googleapis/api-linter/blob/master/rules/aip0158/aip0158.go#L43-L46) function: the message is pagination message only when the message contains the `next_page_token_field` field, or it matches with ("^(List|Search)[A-Za-z0-9]*Response$") pattern. 
2. [isPaginatedRequestMessage](https://github.com/googleapis/api-linter/blob/master/rules/aip0158/aip0158.go#L38-L41) function: the message is pagination message only when the Request message contains `page_size` field or `page_token` field, or it matches with ("^(List|Search)[A-Za-z0-9]*Request$") pattern.

Reply:
Right, I agree with @Dos-Ph here. The reason we attach it to the pagination AIP is because otherwise we have to duplicate the rule for the various different types of methods that might include pagination, which is not what we want.

In fact, I would argue that the non-list methods are actually the most important because they have less defined structure. (Anecdotally, the RPC that made us realize we needed this rule was a custom method.)

Reply:
> The reason we attach it to the pagination AIP is because otherwise we have to duplicate the rule for the various different types of methods that might include pagination, which is not what we want.

I understand that -- and I get that it's easier to say "Paginated stuff should look like this", however we're using pagination as the "easy way to filter things that look like lists and searches". I still argue that we should have a rule saying that even if you didn't do pagination, List and Search responses should have the first item being a repeated field. In other words, I want a rule that will throw an error for:

```
message ListBooksResponse {
  Book books = 1;
}
```

Right now, this rule will *not* do this (correct me if I'm wrong). I suggest we have a simple function that checks:

1. Is this a paginated response (check for next_page_token)? If so, then this message is in play.
2. Is this a standard method that should follow the rule (check for Search*Response and List*Response)? If so, then this message is in play.

If neither of those matches, then continue on. Perhaps this is "IsCollectionResponse" or something similar.

Reply:
I think this should cover your case now.
```
var paginatedReq = regexp.MustCompile("^(List|Search)[A-Za-z0-9]*Request$")
var paginatedRes = regexp.MustCompile("^(List|Search)[A-Za-z0-9]*Response$")

// Return true if this is an AIP-158 List request message, false otherwise.
func isPaginatedRequestMessage(m *desc.MessageDescriptor) bool {
	return paginatedReq.MatchString(m.GetName()) || m.FindFieldByName("page_size") != nil || m.FindFieldByName("page_token") != nil
}

// Return true if this is an AIP-158 List response message, false otherwise.
func isPaginatedResponseMessage(m *desc.MessageDescriptor) bool {
	return paginatedRes.MatchString(m.GetName()) || m.FindFieldByName("next_page_token") != nil
}
```

Reply:
I think we might be mixing two concerns here. There are two things we need:

1. On methods with pagination, we need fields to come in the proper order. That is the domain of this rule.

2. On List and Search methods, we specifically need the first field to be _the resource_, and be plural and repeated. In other words, we need the response message equivalent of [this rule for create](https://linter.aip.dev/133/request-resource-field).

Let's keep these concerns separate. For _this_ rule, let's focus just on the pagination aspect, and I will write a separate rule around making sure the appropriate fields (both resource and pagination) are included in list responses. We have those for most of the rest of the standard methods but this one slipped through the cracks.

Reply:
> On methods with pagination, we need fields to come in the proper order. 
= The first field should be repeated and plural
> On List and Search methods, we specifically need the first field to be the resource, and be plural and repeated
= The first field should be repeated and plural
It looks the same to me.

[Question1] For what you said, you mean that for List and Search response we will have a lint rule for `next_page_token`, right?

[Question2] And for all the response that has `next_page_token` will be considered as Pagination Method. And, we will have lint rules for the following condition, right?
```
The first field in the message (by both position and field number) is not repeated
The first field in the message is a primitive
The first field in the message is not named according to plural(message_name.to_snake().split('_')[1:-1])
```
[Question3] What's about Search Method?

Reply:
> = The first field should be repeated and plural
> Both look the same to me.

There are subtle differences.

- List and Search methods care that the field being repeated over _is the resource being listed_. But a pagination rule does not care what is being paginated.
- The pagination rule cares that the repeated field _lives in a particular spot_ (field 1, at the top of the message), because we do auto-sniffing in some of our tooling to figure out what field to paginate. This is mostly ancillary to list/search; it is specifically a pagination issue.

There is definitely overlap (a field should be repeated and plural) but they are not the same.

> [Question1] For what you said, you mean that for List and Search response we will have a lint rule for `next_page_token`, right? For the List method, we will add the check under AIP-132. [Question2] What's about the Search Method?

Yeah, we need two rules actually, one for each required field. We have them for the request, but not the response -- we just missed them somehow.

We do not have a Search AIP yet, and we need to figure out how strict we want to be on Search. Given the firm rule on having AIP guidance before a lint rule, most likely we would wait on that one.

> [Question3] And for all the response that has `next_page_token` will be considered as Pagination Method. And, we will have lint rules for the following condition defined under AIP-158, right?

Yeah.

Reply:
So, what's about I submit this PR first. And, then create another PR to add next_page_token to List Method lint and update `isPaginatedResponseMessage ` Pagination lint rules?